## Analises dos autos

the below analysis are meant to be an introduction on data cleaning using python and pandas.

It'll be explored the ` eBay Kleinanzeigen` dataset. Ithas the following columns:


`dateCrawled` - When this ad was first crawled. All field-values are taken from his date.

`name` - Name of the car.

`seller` - Whether the seller is private or a dealer.

`offerType` - The type of listing

`price` - The price on the ad to sell the car.

`abtest` - Whether the listing is included in an A/B test.

`vehicleType` - The vehicle Type.

`yearOfRegistration` - The year in which the car was first registered.

`gearbox` - The transmission type.

`powerPS` - The power of the car in PS.

`model` - The car model name.

`kilometer` - How many kilometers the car has driven.

`monthOfRegistration` - The month in which the car was first registered.

`fuelType` - What type of fuel the car uses.

`brand` - The brand of the car.

`notRepairedDamage` - If the car has a damage which is not yet repaired.

`dateCreated` - The date on which the eBay listing was created.

`nrOfPictures` - The number of pictures in the ad.

`postalCode` - The postal code for the location of the vehicle.

`lastSeenOnline` - When the crawler saw this ad last online.



In [34]:
import pandas as pd

autos = pd.read_csv('autos.csv',encoding='Latin-1')

In [47]:
autos.sample(5)

date_crawled                                  name  seller  \
42571  2016-03-28 21:50:40                   Opel_Vechtra_Diesel  privat   
6468   2016-03-20 12:55:51                Mazda_6__Motorschaden_  privat   
35880  2016-03-14 17:38:41  VW_Touran_1.6_FSI_schwarzmetalli_TOP  privat   
11168  2016-04-04 17:37:40                 T4_Caravelle_7_Sitzer  privat   
217    2016-03-25 14:54:53                 PORSCHE_924S__11/1989  privat   

      offerType   price   abtest vehicleType  registration_year  gearbox  \
42571   Angebot    $800  control   limousine               2000  manuell   
6468    Angebot  $1,500     test       kombi               2008  manuell   
35880   Angebot  $6,850     test      andere               2006  manuell   
11168   Angebot  $4,000     test         bus               1999  manuell   
217     Angebot  $8,900  control       coupe               1989      NaN   

       powerPS        model odometer_km  registration_month fuelType  \
42571       74       vectra   150,000km                   3   diesel   
6468       120      6_reihe   150,000km                   4   benzin   
35880      116       touran   150,000km                   8      NaN   
11168      102  transporter   150,000km                   9   diesel   
217        160       andere   150,000km                  11   benzin   

            brand unrepaired_damage           ad_created  nrOfPictures  \
42571        opel              nein  2016-03-28 00:00:00             0   
6468        mazda                ja  2016-03-20 00:00:00             0   
35880  volkswagen              nein  2016-03-14 00:00:00             0   
11168  volkswagen                ja  2016-04-04 00:00:00             0   
217       porsche               NaN  2016-03-25 00:00:00             0   

       postalCode             lastSeen  
42571       48691  2016-03-30 14:16:14  
6468        19258  2016-03-20 12:55:51  
35880       33647  2016-04-02 01:46:42  
11168       39288  2016-04-06 19:17:12  
217         45136  2016-04-06 18:18:44

In [22]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

The dataset has a 50k rows of which the `notRepairedDamage` has the most null rows

In [23]:
autos.rename(columns=dict(yearOfRegistration = 'registration_year',
                         monthOfRegistration = 'registration_month',
                         notRepairedDamage='unrepaired_damage',
                         dateCreated = 'ad_created'),inplace=True)

In [24]:
autos.describe()

registration_year       powerPS  registration_month  nrOfPictures  \
count       50000.000000  50000.000000        50000.000000       50000.0   
mean         2005.073280    116.355920            5.723360           0.0   
std           105.712813    209.216627            3.711984           0.0   
min          1000.000000      0.000000            0.000000           0.0   
25%          1999.000000     70.000000            3.000000           0.0   
50%          2003.000000    105.000000            6.000000           0.0   
75%          2008.000000    150.000000            9.000000           0.0   
max          9999.000000  17700.000000           12.000000           0.0   

         postalCode  
count  50000.000000  
mean   50813.627300  
std    25779.747957  
min     1067.000000  
25%    30451.000000  
50%    49577.000000  
75%    71540.000000  
max    99998.000000

In [25]:
autos['price'] = autos['price'].str.replace(r'[,$]','',regex=True)
autos['price'] = pd.to_numeric(autos['price'])

In [26]:
autos['odometer'] = autos['odometer'].str.replace(r'[,km]','',regex=True)
autos['odometer'] = pd.to_numeric(autos['odometer'])


In [39]:
    autos.rename(columns=dict(yearOfRegistration = 'registration_year',
                            monthOfRegistration = 'registration_month',
                            notRepairedDamage = 'unrepaired_damage',
                            dateCreated = 'ad_created',
                            odometer = 'odometer_km'), inplace=True)

In [28]:
autos['odometer_km'].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [29]:
autos["price"].value_counts()

0        1421
500       781
1500      734
2500      643
1000      639
         ... 
414         1
79933       1
5198        1
18890       1
16995       1
Name: price, Length: 2357, dtype: int64

There's a few rows with price 0, it may be a good ideia to drop them considering it make up only 2% from the dataset

In [30]:
autos["price"].value_counts().sort_index(ascending=False).head(10)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
Name: price, dtype: int64


Aparently, values below 350000 are prone to be rounded and beyond that it jumps to unreasonable values


In [31]:
autos = autos[autos['price'].between(1,350000)]

In [36]:
autos.rename(columns={'dateCrawled':'date_crawled'},
             inplace=True)

In [37]:
autos['date_crawled'].str[:10].value_counts(normalize=True, 
                                            dropna=False)\
                                            .sort_index(ascending=True)

2016-03-05    0.02538
2016-03-06    0.01394
2016-03-07    0.03596
2016-03-08    0.03330
2016-03-09    0.03322
2016-03-10    0.03212
2016-03-11    0.03248
2016-03-12    0.03678
2016-03-13    0.01556
2016-03-14    0.03662
2016-03-15    0.03398
2016-03-16    0.02950
2016-03-17    0.03152
2016-03-18    0.01306
2016-03-19    0.03490
2016-03-20    0.03782
2016-03-21    0.03752
2016-03-22    0.03294
2016-03-23    0.03238
2016-03-24    0.02910
2016-03-25    0.03174
2016-03-26    0.03248
2016-03-27    0.03104
2016-03-28    0.03484
2016-03-29    0.03418
2016-03-30    0.03362
2016-03-31    0.03192
2016-04-01    0.03380
2016-04-02    0.03540
2016-04-03    0.03868
2016-04-04    0.03652
2016-04-05    0.01310
2016-04-06    0.00318
2016-04-07    0.00142
Name: date_crawled, dtype: float64

In [40]:
autos["registration_year"].describe()

count    50000.000000
mean      2005.073280
std        105.712813
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

We find the data `registration_year` to be innacurate, given there are cars registered in year 1000 and 9999, we select a interval to drop all rows between 1900 and 2016. we see below those rows only represents 5% so it is acceptable.

In [45]:
(~autos['registration_year'].between(1900,2016)).sum()/autos.shape[0]

0.03944

In [50]:
autos = autos.query('registration_year >= 1900 and registration_year<=2016')

# Mileage and Price
----

We'll use [pd.query()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.query.html) so we can improve readiblity and performace. We'll be creating a new dataframe with the mean mileage and mean price based on the top brands

In [51]:
brand_counts = autos["brand"].value_counts(normalize=True)
selected_brands = brand_counts[brand_counts > .05].index

mean_mileage = autos.query('brand in @selected_brands')\
    .groupby('brand').mean()['price'].to_dict()

KeyError: 'price'

In [18]:
mean_mileage = autos.query('brand in @selected_brands')\
    .groupby('brand').mean()['price'].to_dict()
mean_mileage = pd.Series(mean_mileage)

In [48]:
mean_prices = autos.groupby('brand')\
                .mean()['price'].to_dict()

mean_prices = pd.Series(mean_prices)

autos_infos = pd.DataFrame(mean_mileage,columns=['mean_mileage'])
autos_infos.insert(1,"mean_prices",mean_prices,True)


KeyError: 'price'